File: backend/lake-matcher/gnis_loader.py
import json
import os
from pathlib import Path
from threading import Lock
from typing import Dict, List

from normalizer import normalize_name

DEFAULT_GNIS_DATA_PATH = Path(__file__).resolve().parent / "data" / "gnis_wa_lakes.json"
HYDROGRAPHY_DATA_PATH = os.getenv("HYDROGRAPHY_DATA_PATH")

_records: List[Dict] = []
_loaded = False
_lock = Lock()


def _load_json(path: Path) -> List[Dict]:
  if not path.exists():
    raise FileNotFoundError(f"GNIS dataset missing at {path}")
  with open(path, "r", encoding="utf-8") as handle:
    return json.load(handle)


def _hydrate(records: List[Dict]) -> List[Dict]:
  hydrated = []
  for record in records:
    normalized = normalize_name(record.get("official_name", ""))
    hydrated.append(
      {
        **record,
        "normalized": normalized,
        "tokens": normalized["tokens"]
      }
    )
  return hydrated


def load_gnis_database() -> List[Dict]:
  global _records, _loaded
  if _loaded:
    return _records

  with _lock:
    if _loaded:
      return _records

    gnis_records = _load_json(Path(os.getenv("GNIS_DATA_PATH", DEFAULT_GNIS_DATA_PATH)))
    hydrated = _hydrate(gnis_records)

    if HYDROGRAPHY_DATA_PATH:
      hydro_path = Path(HYDROGRAPHY_DATA_PATH)
      if hydro_path.exists():
        hydro_records = _hydrate(_load_json(hydro_path))
        hydrated.extend(hydro_records)
      else:
        print(f"[lake-matcher] Hydrography dataset not found at {hydro_path}, ignoring.")

    seen = set()
    deduped: List[Dict] = []
    for record in hydrated:
      key = (record["official_name"], record["latitude"], record["longitude"])
      if key in seen:
        continue
      seen.add(key)
      deduped.append(record)

    _records = deduped
    _loaded = True
    return _records

File: backend/lake-matcher/src/gnisLoader.js
import fs from 'fs';
import path from 'path';
import { DEFAULT_GNIS_DATA_PATH, HYDROGRAPHY_DATA_PATH } from './constants.js';
import { normalizeName } from './normalizer.js';

let combinedRecords = null;
let cachedSignature = null;

function hydrateRecords(records = []) {
  return records.map((record) => ({
    ...record,
    normalized: normalizeName(record.official_name)
  }));
}

function readJsonFile(filePath) {
  if (!fs.existsSync(filePath)) {
    throw new Error(
      `Dataset not found at ${filePath}. Run the GNIS preprocessing script or update GNIS_DATA_PATH.`
    );
  }
  const content = fs.readFileSync(filePath, 'utf8');
  return JSON.parse(content);
}

function loadHydrographyDataset() {
  if (!HYDROGRAPHY_DATA_PATH) {
    return [];
  }
  const resolvedPath = path.resolve(HYDROGRAPHY_DATA_PATH);
  if (!fs.existsSync(resolvedPath)) {
    console.warn(
      `[lake-matcher] Hydrography dataset missing at ${resolvedPath}, continuing with GNIS only.`
    );
    return [];
  }
  const data = readJsonFile(resolvedPath);
  return Array.isArray(data) ? data : [];
}

export function loadGNISDatabase(customPath = DEFAULT_GNIS_DATA_PATH) {
  const resolvedPath = path.resolve(customPath);
  const signature = `${resolvedPath}:${HYDROGRAPHY_DATA_PATH || 'none'}`;
  if (combinedRecords && cachedSignature === signature) {
    return combinedRecords;
  }

  if (!fs.existsSync(resolvedPath)) {
    throw new Error(
      `GNIS dataset was not found at ${resolvedPath}. Run ` +
        '`npm run preprocess:gnis -- <GNIS_Lakes.csv>` from backend/lake-matcher to generate it.'
    );
  }

  const gnisRecords = readJsonFile(resolvedPath);
  const hydratedGnis = hydrateRecords(gnisRecords);
  const hydroRecords = loadHydrographyDataset();
  const hydratedHydro = hydrateRecords(hydroRecords);

  const deduped = new Map();
  for (const record of [...hydratedGnis, ...hydratedHydro]) {
    const key = `${record.official_name}:${record.latitude}:${record.longitude}`;
    if (!deduped.has(key)) {
      deduped.set(key, record);
    }
  }

  combinedRecords = Array.from(deduped.values());
  cachedSignature = signature;
  return combinedRecords;
}
